File: backend/services/model_parameters.py
"""Bounded official-source parameter refresh; no guesses from model names."""
from __future__ import annotations

import json
import math
import os
import re
import tempfile
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any
from urllib.parse import quote

import requests

from backend.config.data_dir import resolve_data_dir
from backend.services.model_parameter_seed import CHECKED_AT, PUBLISHED, UNDISCLOSED

# Only manufacturer-controlled Hugging Face organizations are eligible.
AUTHORS = {
    "alibaba": "Qwen", "google": "google", "mistral": "mistralai",
    "deepseek": "deepseek-ai", "meta": "meta-llama", "nvidia": "nvidia",
    "moonshotai": "moonshotai", "zai": "zai-org", "openai": "openai",
    "allenai": "allenai", "microsoft": "microsoft", "cohere": "CohereForAI",
}
ALIASES = {"qwen": "alibaba", "alibabacloud": "alibaba", "googledeepmind": "google",
           "mistralai": "mistral", "moonshot": "moonshotai", "zhipuai": "zai", "zhipu": "zai"}
_EFFORT = r"\s*\((?:non-reasoning|reasoning|non-thinking|thinking|low|medium|high|xhigh|max)\)$"


def normalized(value: str) -> str:
    return re.sub(r"[^a-z0-9]", "", value.lower())


def creator_key(value: str) -> str:
    value = normalized(value)
    return ALIASES.get(value, value)


def base_name(value: str) -> str:
    return re.sub(_EFFORT, "", value.split("/")[-1], flags=re.I)


def model_key(model: dict[str, Any]) -> str:
    return creator_key(str(model.get("creator", ""))) + ":" + normalized(base_name(str(model.get("name", ""))))


def cache_path() -> Path:
    return resolve_data_dir() / "cache" / "model_parameters.json"


def read_cache(path: Path | None = None) -> dict[str, Any]:
    try:
        value = json.loads((path or cache_path()).read_text(encoding="utf-8"))
        return value if isinstance(value, dict) and isinstance(value.get("entries"), dict) else {"entries": {}}
    except (OSError, ValueError):
        return {"entries": {}}


def write_cache(value: dict[str, Any], path: Path | None = None) -> None:
    target = path or cache_path()
    target.parent.mkdir(parents=True, exist_ok=True)
    from filelock import FileLock
    with FileLock(str(target) + ".lock", timeout=30):
        latest = read_cache(target)
        merged = dict(latest.get("entries", {}))
        for key, entry in value.get("entries", {}).items():
            old = merged.get(key, {})
            # A stale scheduled batch cannot overwrite a later explicit review.
            if str(entry.get("checked_at") or "") >= str(old.get("checked_at") or ""):
                merged[key] = entry
        value = {**value, "entries": merged}
        fd, name = tempfile.mkstemp(prefix=".model-parameters-", dir=target.parent)
        try:
            with os.fdopen(fd, "w", encoding="utf-8") as stream:
                json.dump(value, stream, ensure_ascii=False, allow_nan=False)
            os.replace(name, target)
        finally:
            if os.path.exists(name):
                os.unlink(name)


def valid_counts(value: dict[str, Any]) -> bool:
    total, active = value.get("total"), value.get("active")
    return (isinstance(total, (int, float)) and not isinstance(total, bool)
            and math.isfinite(total) and 0 < total < 1_000_000
            and (active is None or (isinstance(active, (int, float)) and not isinstance(active, bool)
                 and math.isfinite(active) and 0 < active <= total)))


def metadata(model: dict[str, Any], cache: dict[str, Any]) -> dict[str, Any]:
    entry = cache.get("entries", {}).get(model_key(model), {})
    if isinstance(entry, dict) and entry.get("status") == "not_published" and entry.get("source") and entry.get("verification") == "user_review":
        return {k: entry[k] for k in ("status", "source", "checked_at", "verification") if k in entry}
    if isinstance(entry, dict) and entry.get("status") == "known" and valid_counts(entry):
        return {k: entry[k] for k in ("status", "total", "active", "source", "checked_at", "verification") if k in entry}
    names = {normalized(base_name(str(model.get(field, "")))) for field in ("name", "slug")}
    creator = creator_key(str(model.get("creator", "")))
    for seed in PUBLISHED:
        if creator_key(seed["creator"]) == creator and any(normalized(n) in names for n in seed["names"]):
            return {"status": "known", "checked_at": CHECKED_AT,
                    **{k: seed[k] for k in ("total", "active", "source") if k in seed}}
    closed_name = normalized(re.sub(r"\s*\([^)]*\)$", "", str(model.get("name", ""))))
    for disclosure in UNDISCLOSED:
        if creator_key(disclosure["creator"]) == creator and closed_name in map(normalized, disclosure["names"]):
            return {"status": "not_published", "source": disclosure["source"], "checked_at": CHECKED_AT}
    return {"status": "pending"}


def enrich_comparison(feed: dict[str, Any]) -> dict[str, Any]:
    cache = read_cache()
    return {**feed, "models": [{**model, "parameter_metadata": metadata(model, cache)}
                              for model in feed.get("models", [])]}


def _billions(number: str, unit: str) -> float:
    return float(number.replace(",", ".")) * {"m": .001, "b": 1, "t": 1000}[unit.lower()[0]]


def parse_card(text: str, repo: str) -> dict[str, float] | None:
    """Only explicit count fields or the exact model's table row, not benchmarks."""
    text = text.replace("**", "").replace("__", "")
    number = r"([0-9]+(?:[.,][0-9]+)?)\s*(B|M|T|billion|million|trillion)\b"
    result: dict[str, float] = {}
    match = re.search(r"(?:Number of Parameters|Total Parameters)\s*[:|]\s*" + number, text, re.I)
    if match:
        result["total"] = _billions(match[1], match[2])
        line = text[match.start():].splitlines()[0]
        active = re.search(number + r"\s*(?:activated|active)", line, re.I)
        if active:
            result["active"] = _billions(active[1], active[2])
        active = re.search(r"(?:Activated|Active) Parameters\s*[:|]\s*" + number, text, re.I)
        if active:
            result["active"] = _billions(active[1], active[2])
    if not result:
        for line in text.splitlines():
            cells = [cell.strip() for cell in line.strip().strip("|").split("|")]
            if len(cells) < 3 or normalized(cells[0]) != normalized(repo.split("/")[-1]):
                continue
            total = re.fullmatch(number, cells[1], re.I)
            active = re.fullmatch(number, cells[2], re.I)
            if total and active:
                result = {"total": _billions(total[1], total[2]), "active": _billions(active[1], active[2])}
    return result if valid_counts(result) else None


def _get(url: str, *, params: dict[str, Any] | None = None) -> bytes:
    # URLs are constructed here from pinned official authors, never supplied by cards.
    with requests.get(url, params=params, timeout=(4, 8), allow_redirects=False, stream=True) as response:
        response.raise_for_status()
        if response.status_code != 200:
            raise ValueError("Unexpected source response")
        chunks, size = [], 0
        for chunk in response.iter_content(8192):
            size += len(chunk)
            if size > 2_000_000:
                raise ValueError("Official model card exceeds size limit")
            chunks.append(chunk)
        return b"".join(chunks)


def _identity(value: str) -> str:
    # Instruction-tuning suffixes are allowed; version/date/size suffixes are preserved.
    return normalized(re.sub(r"(?:[- ](?:instruct|it))$", "", base_name(value), flags=re.I))


def discover(model: dict[str, Any], author: str) -> dict[str, Any] | None:
    name = base_name(str(model.get("name", "")))
    candidates = json.loads(_get("https://huggingface.co/api/models", params={
        "author": author, "search": re.sub(r"\s+", "-", name), "limit": 20,
    }))
    identities = {_identity(str(model.get(field, ""))) for field in ("name", "slug")}
    matches = []
    for candidate in candidates if isinstance(candidates, list) else []:
        if not isinstance(candidate, dict):
            continue
        repo = str(candidate.get("id", ""))
        if (re.fullmatch(r"[\w.-]+/[\w.-]+", repo) and repo.split("/")[0] == author
                and _identity(repo) in identities):
            matches.append(repo)
    # Ambiguous versions require review; never select the first similar result.
    if len(matches) != 1:
        return None
    repo = matches[0]
    card = _get(f"https://huggingface.co/{quote(repo, safe='/')}/raw/main/README.md").decode("utf-8")
    counts = parse_card(card, repo)
    return {**counts, "source": f"https://huggingface.co/{repo}"} if counts else None


def refresh_parameters(*, models: list[dict[str, Any]] | None = None,
                       path: Path | None = None, max_models: int = 40,
                       budget_seconds: float = 120) -> dict[str, Any]:
    if models is None:
        from backend.services.artificial_analysis import _read_cache, fetch_all_models
        feed = _read_cache() or fetch_all_models()
        models = feed.get("models", [])
    cache = read_cache(path)
    entries = cache["entries"]
    unique = {model_key(model): model for model in models}
    now = datetime.now(timezone.utc).isoformat()
    deadline = time.monotonic() + budget_seconds
    attempted = updated = failed = 0
    ordered = sorted(unique)
    cursor = cache.get("cursor", "")
    ordered = [key for key in ordered if key > cursor] + [key for key in ordered if key <= cursor]
    for key in ordered:
        if attempted >= max_models or time.monotonic() >= deadline:
            break
        model = unique[key]
        author = AUTHORS.get(creator_key(str(model.get("creator", ""))))
        if not author:
            continue
        attempted += 1
        cache["cursor"] = key
        previous = metadata(model, cache)
        try:
            found = discover(model, author)
            if found:
                entries[key] = {**found, "status": "known", "checked_at": now}
                updated += 1
            else:
                entries[key] = {**previous, "attempted_at": now, "reason": "needs_review"}
        except (requests.RequestException, ValueError, KeyError, TypeError):
            failed += 1
            entries[key] = {**previous, "attempted_at": now, "reason": "source_unavailable"}
    cache["last_run"] = now
    write_cache(cache, path)
    return {"success": failed < attempted or attempted == 0, "attempted": attempted,
            "verified": updated, "source_errors": failed,
            "pending": sum(metadata(model, cache)["status"] == "pending" for model in unique.values()),
            "message": "Official parameter sources checked; unmatched models remain pending review."}

File: frontend/src/features/settings/model-comparison/modelParameters.ts
import type { AiModelComparisonEntry } from '../../../shared/api/ai';

interface ParameterMetadata {
    /** Billions of parameters (10^9), not file size or RAM. */
    readonly total: number;
    readonly active?: number;
    readonly source: string;
    readonly checkedAt: string;
    readonly verification?: string;
}

interface PublishedParameters extends Omit<ParameterMetadata, 'checkedAt'> {
    readonly creator: string;
    readonly names: readonly string[];
}

// Manufacturer disclosures checked on 2026-09-19. Exact aliases only: never
// infer a count from a model name, nor propagate one to a later generation.
const PUBLISHED_PARAMETERS: readonly PublishedParameters[] = [
    { creator: 'Alibaba', names: ['Qwen3-4B', 'Qwen3 4B'], total: 4,
        source: 'https://huggingface.co/Qwen/Qwen3-4B' },
    { creator: 'Alibaba', names: ['Qwen3-8B', 'Qwen3 8B'], total: 8.2,
        source: 'https://huggingface.co/Qwen/Qwen3-8B' },
    { creator: 'Alibaba', names: ['Qwen3-14B', 'Qwen3 14B'], total: 14.8,
        source: 'https://huggingface.co/Qwen/Qwen3-14B' },
    { creator: 'Alibaba', names: ['Qwen3-32B', 'Qwen3 32B'], total: 32.8,
        source: 'https://huggingface.co/Qwen/Qwen3-32B' },
    { creator: 'Alibaba', names: ['Qwen3-30B-A3B', 'Qwen3 30B A3B'], total: 30.5, active: 3.3,
        source: 'https://huggingface.co/Qwen/Qwen3-30B-A3B' },
    { creator: 'Alibaba', names: ['Qwen3-235B-A22B-Instruct-2507', 'Qwen3 235B A22B Instruct 2507'], total: 235, active: 22,
        source: 'https://huggingface.co/Qwen/Qwen3-235B-A22B-Instruct-2507' },
    { creator: 'Alibaba', names: ['Qwen3-235B-A22B-Thinking-2507', 'Qwen3 235B A22B Thinking 2507'], total: 235, active: 22,
        source: 'https://huggingface.co/Qwen/Qwen3-235B-A22B-Thinking-2507' },
    { creator: 'Alibaba', names: ['Qwen3-30B-A3B-Instruct-2507', 'Qwen3 30B A3B Instruct 2507'], total: 30.5, active: 3.3,
        source: 'https://huggingface.co/Qwen/Qwen3-30B-A3B-Instruct-2507' },
    { creator: 'Alibaba', names: ['Qwen3-30B-A3B-Thinking-2507', 'Qwen3 30B A3B Thinking 2507'], total: 30.5, active: 3.3,
        source: 'https://huggingface.co/Qwen/Qwen3-30B-A3B-Thinking-2507' },
    { creator: 'Alibaba', names: ['Qwen3-Coder-480B-A35B-Instruct', 'Qwen3 Coder 480B A35B', 'Qwen3 Coder 480B A35B Instruct'], total: 480, active: 35,
        source: 'https://huggingface.co/Qwen/Qwen3-Coder-480B-A35B-Instruct' },
    { creator: 'Alibaba', names: ['Qwen3-Coder-30B-A3B-Instruct', 'Qwen3 Coder 30B A3B', 'Qwen3 Coder 30B A3B Instruct'], total: 30.5, active: 3.3,
        source: 'https://huggingface.co/Qwen/Qwen3-Coder-30B-A3B-Instruct' },
    { creator: 'Alibaba', names: ['QwQ-32B', 'QwQ 32B'], total: 32.5,
        source: 'https://huggingface.co/Qwen/QwQ-32B' },
    { creator: 'Google', names: ['Gemma 3 1B', 'Gemma 3 1B Instruct', 'gemma-3-1b-it'], total: 1,
        source: 'https://ai.google.dev/gemma/docs/core/model_card_3' },
    { creator: 'Google', names: ['Gemma 3 4B', 'Gemma 3 4B Instruct', 'gemma-3-4b-it'], total: 4,
        source: 'https://ai.google.dev/gemma/docs/core/model_card_3' },
    { creator: 'Google', names: ['Gemma 3 12B', 'Gemma 3 12B Instruct', 'gemma-3-12b-it'], total: 12,
        source: 'https://ai.google.dev/gemma/docs/core/model_card_3' },
    { creator: 'Google', names: ['Gemma 3 27B', 'Gemma 3 27B Instruct', 'gemma-3-27b-it'], total: 27,
        source: 'https://ai.google.dev/gemma/docs/core/model_card_3' },
    { creator: 'Mistral', names: ['Mistral-Small-24B-Instruct-2501', 'Mistral Small 3'], total: 24,
        source: 'https://huggingface.co/mistralai/Mistral-Small-24B-Instruct-2501' },
    { creator: 'Mistral', names: ['Mistral-Small-3.1-24B-Instruct-2503', 'Mistral Small 3.1'], total: 24,
        source: 'https://huggingface.co/mistralai/Mistral-Small-3.1-24B-Instruct-2503' },
    { creator: 'Mistral', names: ['Mistral-Small-3.2-24B-Instruct-2506', 'Mistral Small 3.2'], total: 24,
        source: 'https://huggingface.co/mistralai/Mistral-Small-3.2-24B-Instruct-2506' },
    { creator: 'DeepSeek', names: ['DeepSeek-R1', 'DeepSeek R1', "DeepSeek R1 (Jan '25)"], total: 671, active: 37,
        source: 'https://huggingface.co/deepseek-ai/DeepSeek-R1' },
    { creator: 'DeepSeek', names: ['DeepSeek-V3.1', 'DeepSeek V3.1'], total: 671, active: 37,
        source: 'https://huggingface.co/deepseek-ai/DeepSeek-V3.1' },
    { creator: 'NVIDIA', names: ['NVIDIA-Nemotron-3-Nano-30B-A3B-BF16', 'NVIDIA Nemotron 3 Nano 30B A3B', 'Nemotron 3 Nano 30B A3B'], total: 30, active: 3.5,
        source: 'https://huggingface.co/nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16' },
    { creator: 'Moonshot AI', names: ['Kimi-K2-Instruct', 'Kimi K2', 'Kimi K2 Instruct'], total: 1000, active: 32,
        source: 'https://huggingface.co/moonshotai/Kimi-K2-Instruct' },
    { creator: 'Moonshot AI', names: ['Kimi-K2-Thinking', 'Kimi K2 Thinking'], total: 1000, active: 32,
        source: 'https://huggingface.co/moonshotai/Kimi-K2-Thinking' },
    { creator: 'Z AI', names: ['GLM-4.5', 'GLM 4.5'], total: 355, active: 32,
        source: 'https://huggingface.co/zai-org/GLM-4.5' },
    { creator: 'Z AI', names: ['GLM-4.5-Air', 'GLM 4.5 Air'], total: 106, active: 12,
        source: 'https://huggingface.co/zai-org/GLM-4.5-Air' },

    { creator: 'OpenAI', names: ['gpt-oss-120b'], total: 117, active: 5.1,
        source: 'https://openai.com/index/introducing-gpt-oss/' },
    { creator: 'OpenAI', names: ['gpt-oss-20b'], total: 21, active: 3.6,
        source: 'https://openai.com/index/introducing-gpt-oss/' },
    { creator: 'Mistral', names: ['Devstral 2', 'Devstral-2-123B-Instruct-2512'], total: 123,
        source: 'https://mistral.ai/news/devstral-2-vibe-cli/' },
    { creator: 'Mistral', names: ['Devstral Small 2', 'Devstral-Small-2-24B-Instruct-2512'], total: 24,
        source: 'https://mistral.ai/news/devstral-2-vibe-cli/' },
    { creator: 'DeepSeek', names: ["DeepSeek V3 (Dec '24)", 'DeepSeek-V3'], total: 671, active: 37,
        source: 'https://huggingface.co/deepseek-ai/DeepSeek-V3' },
    { creator: 'Alibaba', names: ['Qwen3 235B A22B', 'Qwen3-235B-A22B'], total: 235, active: 22,
        source: 'https://huggingface.co/Qwen/Qwen3-235B-A22B' },
    { creator: 'Meta', names: ['Llama 4 Maverick', 'Llama-4-Maverick-17B-128E-Instruct'], total: 400, active: 17,
        source: 'https://huggingface.co/meta-llama/Llama-4-Maverick-17B-128E-Instruct' },
    { creator: 'Meta', names: ['Llama 4 Scout', 'Llama-4-Scout-17B-16E-Instruct'], total: 109, active: 17,
        source: 'https://huggingface.co/meta-llama/Llama-4-Scout-17B-16E-Instruct' },
];

const CHECKED_AT = '2026-09-19';

const normalize = (value: string): string => value.toLowerCase().replace(/[^a-z0-9]/g, '');
const baseName = (value: string): string => (value.split('/').pop() ?? value).replace(
    /\s*\((?:non-reasoning|reasoning|non-thinking|thinking|low|medium|high|xhigh|max)\)$/i, '',
);

const creatorKey = (value: string): string => {
    const key = normalize(value);
    return ({ qwen: 'alibaba', alibabacloud: 'alibaba', mistralai: 'mistral',
        moonshot: 'moonshotai', zai: 'zai', zhipuai: 'zai', zhipu: 'zai',
        googledeepmind: 'google' } as Record<string, string>)[key] ?? key;
};

export function modelParameterMetadata(model: AiModelComparisonEntry): ParameterMetadata | null {
    const remote = model.parameter_metadata;
    if (remote?.status === 'known' && typeof remote.total === 'number' && remote.source) {
        return { total: remote.total, ...(typeof remote.active === 'number' ? { active: remote.active } : {}),
            source: remote.source, checkedAt: remote.checked_at ?? '', verification: remote.verification ?? undefined };
    }

    if (remote?.status === 'not_published') return null;
    const names = [model.name, model.slug].map((name) => normalize(baseName(name)));
    const entry = PUBLISHED_PARAMETERS.find((entry) => (
        creatorKey(entry.creator) === creatorKey(model.creator)
        && entry.names.some((name) => names.includes(normalize(name)))
    ));
    return entry ? { ...entry, checkedAt: CHECKED_AT } : null;
}

// These specific model families were reviewed in the official specifications.
// "Not published" means absent from that documentation on checkedAt, not a
// permanent claim about a vendor. Unreviewed releases remain pending.
const UNDISCLOSED_PARAMETERS = [
    { creator: 'Anthropic', names: ['Claude Fable 5.1', 'Claude Opus 5', 'Claude Sonnet 5', 'Claude Haiku 4.5'],
        source: 'https://platform.claude.com/docs/en/models/overview' },
    { creator: 'OpenAI', names: ['GPT-6 Astra', 'GPT-5.6 Sol', 'GPT-5.6 Terra', 'GPT-5.6 Luna', 'GPT-5.5', 'GPT-4o', 'GPT-4o mini', 'GPT-4.1', 'GPT-4.1 mini', 'GPT-4.1 nano', 'o3', 'o3-pro', 'o4-mini'],
        source: 'https://developers.openai.com/api/docs/models' },
    { creator: 'Google', names: ['Gemini 3.8 Flash', 'Gemini 3.7 Flash', 'Gemini 3.6 Flash', 'Gemini 3.5 Flash', 'Gemini 3.5 Flash-Lite', 'Gemini 3.1 Flash-Lite'],
        source: 'https://ai.google.dev/gemini-api/docs/models' },
] as const;

export function modelParameterDisclosure(model: AiModelComparisonEntry): {
    status: 'not_published' | 'pending'; source?: string; checkedAt?: string;
} {

    const remote = model.parameter_metadata;
    if (remote?.status === 'not_published') {
        return { status: 'not_published', source: remote.source ?? undefined, checkedAt: remote.checked_at ?? undefined };
    }
    // Effort/fallback settings do not change the reviewed closed model family.
    const name = normalize(model.name.replace(/\s*\([^)]*\)$/, ''));
    const entry = UNDISCLOSED_PARAMETERS.find((candidate) => (
        creatorKey(candidate.creator) === creatorKey(model.creator)
        && candidate.names.some((alias) => normalize(alias) === name)
    ));
    return entry ? { status: 'not_published', source: entry.source, checkedAt: CHECKED_AT }
        : { status: 'pending' };
}
